pkg/geo/geomfn/generate_points.go

```diff
@@ -0,0 +1,227 @@
+	}
+	return out, nil
+}
+
+func polygonGeometryToPoints(g geo.Geometry, npoints int, seed int64) (*geom.MultiPoint, error) {
+	area, err := Area(g)
+	if err != nil {
+		return nil, errors.Newf("could not calculate Polygon area: %v", err)
+	}
+	bbox := g.CartesianBoundingBox()
+	bboxWidth := bbox.HiX - bbox.LoX
+	bboxHeight := bbox.HiY - bbox.LoY
+	bboxArea := bboxHeight * bboxWidth
+	if area == 0.0 || bboxArea == 0.0 {
+		return nil, errors.New("zero area input Polygon")
+	}
+	// Gross up our test set a bit to increase odds of getting coverage in one pass.
+	sampleNpoints := float64(npoints) * bboxArea / area
+
+	sampleSqrt := math.Round(math.Sqrt(sampleNpoints))
+	if sampleSqrt == 0 {
+		sampleSqrt = 1
+	}
+	var sampleHeight, sampleWidth int
+	var sampleCellSize float64
+	// Calculate the grids we're going to randomize within.
+	if bboxWidth > bboxHeight {
+		sampleWidth = int(sampleSqrt)
+		sampleHeight = int(math.Ceil(sampleNpoints / float64(sampleWidth)))
+		sampleCellSize = bboxWidth / float64(sampleWidth)
+	} else {
+		sampleHeight = int(sampleSqrt)
+		sampleWidth = int(math.Ceil(sampleNpoints / float64(sampleHeight)))
+		sampleCellSize = bboxHeight / float64(sampleHeight)
+	}
+	// Prepare the polygon for fast true/false testing.
+	gprep, err := geos.PrepareGeometry(g.EWKB())
+	if err != nil {
+		return nil, errors.Newf("could not prepare Geometry: %v", err)
+	}
+	defer func() {
+		err = geos.PreparedGeomDestroy(gprep)
+	}()
+
+	// Initiate random number generator.
+	randomNumberGenerator := rand.New(rand.NewSource(seed))
+	// Generate a slice of points - for every cell on a grid store coordinates.
+	n := sampleHeight * sampleWidth
+	cells := make([]geom.Coord, n)
+	for i := 0; i < sampleWidth; i++ {
+		for j := 0; j < sampleHeight; j++ {
+			cells[i*int(sampleHeight)+j] = geom.Coord{float64(i), float64(j)}
+		}
+	}
+	// Shuffle the points. Without shuffling, the generated point will
+	// always be adjacent to the previous one (in terms of grid cells).
+	if n > 1 {
+		randomNumberGenerator.Shuffle(n, func(i int, j int) {
+			temp := cells[j]
+			cells[j] = cells[i]
+			cells[i] = temp
+		})
+	}
+	results := geom.NewMultiPoint(geom.XY)
+	// Generate points and test them.
+	npointsGenerated := 0
+	iterations := 0
+	for npointsGenerated < npoints {
+		iterations++
+		for _, cell := range cells {
+			y := bbox.LoY + cell.X()*sampleCellSize
+			x := bbox.LoX + cell.Y()*sampleCellSize
+			x += randomNumberGenerator.Float64() * sampleCellSize
+			y += randomNumberGenerator.Float64() * sampleCellSize
+			if x > bbox.HiX || y > bbox.HiY {
+				continue
+			}
+			gseq, err := geos.CoordSequenceCreate(1, 2)
+			if err != nil {
+				return nil, errors.Newf("could not create coord sequence: %v", err)
+			}
+			gseq, err = geos.CoordSequenceSetXY(gseq, 0, x, y)
+			if err != nil {
+				return nil, errors.Newf("could not set X and Y for coord sequence: %v", err)
+			}
+			defer func() {
+				// err = geos.CoordSequenceDestroy(gseq)
```

Comment:
1) I know that `gseq` should be destroyed, but when I uncomment this line I get a really weird error and can't figure out why:
```
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x5fd1210]

runtime stack:
runtime.throw(0x4a2e35f, 0x2a)
	/usr/local/Cellar/go/1.15.5/libexec/src/runtime/panic.go:1116 +0x72
runtime.sigpanic()
	/usr/local/Cellar/go/1.15.5/libexec/src/runtime/signal_unix.go:726 +0x48c
```
I compared the C API handler code to the one for `PreparedGeometry`, I have noticed that the destroy method for `CoordSequence` has no `const` argument type:
```
extern void GEOS_DLL GEOSPreparedGeom_destroy_r(GEOSContextHandle_t handle,
                                                const GEOSPreparedGeometry* g);

extern void GEOS_DLL GEOSCoordSeq_destroy_r(GEOSContextHandle_t handle,
                                            GEOSCoordSequence* s);
```
Could you help to debug it, please? 🙏 
2) Can you recommend any better way to handle errors in the `defer` statement?

Reply:
Still on (1) - sorry, but how do I do that? I have tried to add a new function like this:
```
CR_GEOS_Status CR_GEOS_GeometryDestroy(CR_GEOS* lib, CR_GEOS_Slice a) {
  std::string error;
  auto handle = initHandleWithErrorBuffer(lib, &error);

  auto wkbReader = lib->GEOSWKBReader_create_r(handle);
  auto geomA = lib->GEOSWKBReader_read_r(handle, wkbReader, a.data, a.len);
  lib->GEOSWKBReader_destroy_r(handle, wkbReader);
  if (geomA != nullptr) {
    lib->GEOSGeom_destroy_r(handle, geomA);
  }
  lib->GEOS_finish_r(handle);
  return toGEOSString(error.data(), error.length());
}
```
But in the end, I still got the same error.
I could make this function expect a `CR_GEOS_Geometry` type input but in that case, I would probably need another way of keeping a pointer to `CR_GEOS_Geometry` GeometryPoint and transforming it to a `CR_GEOS_Slice` (so that it can be used in `PrepareGeometry`).

Reply:
I think you don't need to call `GEOSCoordSeq_destroy_r` at all, since `GeometryCreate` makes it the owner of the Seq object.

I think we should avoid doing this GeometryCreate altogether. Can we just create a Point using `go-geom` and use that EWKB?

Also for simplicity, I'm happy for you to just use `geos.Intersects` instead of `PreparedGeometry`.

Reply:
Ah, that makes sense. I will go with a simple point then 👍